### DISCUSSION TITLE
How to use Rust and Netbeans?

### DISCUSSION THREAD

**Discussion by jcbritobr** [apache/netbeans#6264] [General]
Read that netbeans 18 has initial support for rust, but nothing in the application shows support for it, and there is no documentation. How to load a rust project?

**Comment by troizet** [comment]
Try building Netbeans from source with the command `$ ant -Dcluster.config=full build`

**Reply by mbien** [reply]
^^ is correct. The rust module cluster isn't part of the release build yet. It is still baking. `-Dcluster.config=full` will tell the build to include everything. You could also set it to `rust` to build a rust IDE.

see readme: https://github.com/apache/netbeans#building-netbeans

**Reply by mwoodiupui** [reply]
Two years later, it's still not part of the v26 release build.  Everything on the web points to an 8-year-old plugin that won't build.  After you fix `http:` -> `https:` in the POM, Maven says the repo. is closed and points you to an "Oracle transition" document that no longer exists.

**Reply by matthiasblaesing** [reply]
@mwoodiupui read the first entry in this thread. The rust support is part of the NetBeans built.

**Reply by Chris2011** [reply]
To be honest, what is the reason after 2/3 years now, to not include it fully into it without building the full cluster manually. @vieiro what is missing here, that we are afraid of not putting it into? New tickets about missing stuff and bugs?

**Reply by matthiasblaesing** [reply]
Anyone interested in rust on NetBeans might be interested in this Branch:

https://github.com/matthiasblaesing/netbeans/tree/rust-analyser

The idea is to use rust-analyer (the Rust LSP server) to provide the required services. I'm waiting for an updated toml-java and will then create a PR from that.

The idea of that branch is, that the grammar based analysis remains in place as fallback, but user in encouraged to use rust-analyzer.

Still missing:
- once per session info that rust analyzer should be downloaded for better support
- alignment of icons between lsp-client and rust.grammar if possible.

If needed I can provide a prebuilt binary for the branch.

**Reply by Chris2011** [reply]
So this is like a hybrid if the LSP breaks somehow, it goes back to the "traditional" lexer/parser stuff, right?

**Reply by matthiasblaesing** [reply]
"LSP breaks" in this case means "user did not configure rust-analyzer". For Typescript the situation is mostly trivial: Noone develops typescript without having node installed, so we can carry the TS LSP. For Rust the LSP is platform dependent so the user has to download a matching version. The binary is 50MB.

**Reply by vieiro** [reply]
@matthiasblaesing is right, `rust-analyzer` is the way to go. It's very easy to install: 

```
$ rustup component add rust-analyzer
```
(which can be ran from within the cluster in case of need).

Some stuff needing care in the `rust` cluster, additionally to Matthias' plan:

- Rust grammar needs improvements (e.g. declaring variables in multiple parts of a function).
- LSP can do a good job in indenting, but some default indenting should be in plance, I think.

I have seen little interest in `rust` support over the years, and I haven't find time to dedicate to it. But I may be able to help in specific parts if required.

I'll take a look at the `toml` enhancements during the next two weeks.

**Reply by Chris2011** [reply]
> I have seen little interest in rust support over the years,

Honestly I can understand this, but I see it with the thing that we had of a dedicated kotlin plugin and the long discussion in the ticket: "We cancled the plugin, because little interest" and users: "We can't use it, because it is buggy". Don't get me wrong, it is not about buggy here, but for "normal" users/consumers, they don't want to build the IDE and I guess those are more than we expect.

Me either, I just want to download the latest IDE and use it

**Comment by matthiasblaesing** [comment]
So I brought my branch into a state where I'm willing to get it into NetBeans: https://github.com/apache/netbeans/pull/9249

That does not yet enable the rust cluster by default but I'm inclined to do that.
